### DISCUSSION TITLE
Mass-edit/ Bulk-edit / Quick-edit /Default policies discussion thread

### DISCUSSION THREAD

**Discussion by kianzarrin** [CitiesSkylinesMods/TMPE#1666] [General]
The following issues 
#148 #542 #541 #539 #540 #523 #537 #557
discuss how to mass edit rues in different circumstances. I create a this thread for a general discussion about mass edits.
First I like to distinguish between these concepts (and I am also introducing terminology I hope everyone likes it :) ) : 
 A. Default policy: Rules that apply by default to all of the city. see screenshot 1. 
 B. Quick edit: quickly apply commonly used road rules to a junction/road/roundabout such as  #542 #541 #539 #540. (the user have some control over this in the options)
 C. Bulk edit: The user can select multiple segments/junctions and apply traffic rules. This takes more time than quick edit but gives user more flexibility. This is useful for cases that are not so common (or we have not thought of). (Issue ~~TBC~~ #560) .
 D. Mass edit: a general term for Bulk-edit or Quick edit.

![Screenshot (75)](https://user-images.githubusercontent.com/26344691/69384873-375b2700-0cc6-11ea-9ad6-dcb4b239fde3.png)

Currently #541 #539 are done by quick edit. but in future if possible I want to make it default policy so that the user does not even have do ctrl+shift+click create priority road or fix roundabout rules.  I am not sure if this is possible.

Another improvement is that the policies could be district specific. (issue ~~TBC~~ #561). 

The 'Adjust road selection tool' is currently being  used as quick-edit in #542. If we can remember the selection then we can automatically apply rules when user attaches new roads to the priority road/roundabout (issue ~~TBC~~ #557).

as for bulk edit the user may select different segments/junctions by ALT+SHIFT+CLICKing (or any other combination of modifier keys we may decide ) on them. The 'Adjust road selection toot'  (currently begin used for quick-edit) has good potential for being used as a tool to select a tool to select multiple road segments for bulk edit.

EDIT: created #568 regarding undoing mass edits.

**Comment by originalfoo** [comment]
We should be careful when auto-applying stuff by default; ideally user should always trigger the changes manually (including manual use of bulk application tools). See: https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/issues/541#issuecomment-557320091

**Comment by kianzarrin** [comment]
@aubergine10 
https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/issues/541#issuecomment-557320091
>Also note that there is a 16 MB limit (link to old repo) to the amount of data TM:PE can store in save games, and automatically applying customization could potentially push over that limit on large cities; and even if not, might cause general performance issues.

We already have default policies in the options policy tab. And they do not take more than a few bytes to save/load. Setting up default policies does not need any storage. We don't need to save anything if we know they apply to the whole city. We only need to store information if the user override the defaults.

>This could be very confusing for a lot of users. For example, we get regular issues where people have set up crazy customisations and then wonder why their traffic is broken. So I think it best to avoid automatically applying customisations; if user has to manually apply (including manual bulk application) and something goes wrong, at least they have some idea of what they did to make it go wrong.

That's true. But sometimes default policies can be useful such as:
- most of the ones we already have
- When we set a priority road in the `adjust road selection tool` (#542 ) We can automatically apply traffic rules to all new small roads that may be connected to the priority road/roundabout. although this blurs the line between what we consider quick-edit and default policy.
- when a small road is connected to a main road, simcity only allows right turn and puts a stop sign. We can have an option to do the same in CS-TMPE . To avoid confusion I do not recommend to do this until we have a way I putting continues junction barrier to provide visual confirmation to the user as to what is going on like this picture:
![68045810-0e58ff00-fce3-11e9-85c5-bf12748d6f36](https://user-images.githubusercontent.com/26344691/69388546-d33e6000-0cd1-11ea-9a5a-5a015c5bf2c3.png)
- I cannot see any harm in using default settings on circular roundabouts. The user can override them if he wants.
- when an avenue is split in two like this picture:
![Screenshot (81)](https://user-images.githubusercontent.com/26344691/69388678-6b3c4980-0cd2-11ea-9238-20e5a88ad32a.png)

**Comment by originalfoo** [comment]
There are some highway roads that work similar to avenues (single road with both directions), for example [NExt2](https://steamcommunity.com/sharedfiles/filedetails/?id=812125426) and [UK Motorways](https://steamcommunity.com/sharedfiles/filedetails/?id=1718180429) - will they work the same as avenue above if split in to two one-way roads?

**Comment by kvakvs** [comment]
The user needs to know:
* that feature exists
* how to enable the feature
* what it does
* whether it actually worked and how to revert it (redo some other way)

P.S. Change notes are not a good place where you tell about the feature :) It should be in game UI somehow, in a tooltip, help, or naturally offered via visual hints.

**Comment by originalfoo** [comment]
^ that comment is going in the UI style guide!

**Comment by kianzarrin** [comment]
> The user needs to know:
> 
> * that feature exists
> * how to enable the feature
> * what it does
> * whether it actually worked and how to revert it (redo some other way)
> 
> P.S. Change notes are not a good place where you tell about the feature :) It should be in game UI somehow, in a tooltip, help, or naturally offered via visual hints.

I should put wiki help and advisor. 
I cannot think of an easy way to revert the changes unless if we provide CTRL+z option.

**Comment by originalfoo** [comment]
Tagging #854 , #853
